crates/unftp-sbe-fs/examples/cap-ftpd-worker.rs

```diff
@@ -0,0 +1,212 @@
+        async fn authenticate(&self, username: &str, creds: &libunftp::auth::Credentials) -> Result<User, AuthenticationError> {
+            let res = if let Some(actual_creds) = self.credentials_map.get(username) {
+                let pass_check_result = match &creds.password {
+                    Some(ref given_password) => {
+                        if Self::check_password(given_password, &actual_creds.password).is_ok() {
+                            Some(Ok(User::new(username, &actual_creds.home)))
+                        } else {
+                            Some(Err(AuthenticationError::BadPassword))
+                        }
+                    }
+                    None => None,
+                };
+
+                match pass_check_result {
+                    None => Err(AuthenticationError::BadPassword),
+                    Some(pass_res) => {
+                        if pass_res.is_ok() {
+                            Ok(User::new(username, &actual_creds.home))
+                        } else {
+                            pass_res
+                        }
+                    }
+                }
+            } else {
+                Err(AuthenticationError::BadUser)
+            };
+
+            if res.is_err() {
+                sleep(Duration::from_millis(1500)).await;
+            }
+
+            res
+        }
+
+        fn name(&self) -> &str {
+            std::any::type_name::<Self>()
+        }
+    }
+
+    #[async_trait]
+    impl Authenticator<DefaultUser> for JsonFileAuthenticator {
+        #[tracing_attributes::instrument]
+        async fn authenticate(&self, username: &str, creds: &libunftp::auth::Credentials) -> Result<DefaultUser, AuthenticationError> {
+            let _: User = self.authenticate(username, creds).await?;
+            Ok(DefaultUser {})
+        }
+    }
+}
+
+use auth::{JsonFileAuthenticator, User};
+
+#[tokio::main(flavor = "current_thread")]
+async fn main() {
+    println!("Starting helper");
+    let args: Vec<String> = env::args().collect();
+
+    if args.len() != 3 {
+        eprintln!("Usage: {} <AUTH_FILE> <FD>", args[0]);
+        exit(2);
+    }
+    let fd: RawFd = if let Ok(fd) = args[2].parse() {
+        fd
+    } else {
+        eprintln!("Usage: {} <FD>\nFD must be numeric", args[0]);
+        exit(2)
+    };
+
+    let std_stream = unsafe { std::net::TcpStream::from_raw_fd(fd) };
+
+    let control_sock = TcpStream::from_std(std_stream).unwrap();
+
+    let auth = Arc::new(JsonFileAuthenticator::from_file(args[1].clone()).unwrap());
+    // XXX This would be a lot easier if the libunftp API allowed creating the
+    // storage just before calling service.
+    let storage = Mutex::new(Some(Filesystem::new(std::env::temp_dir())));
+    let sgen = Box::new(move || storage.lock().unwrap().take().unwrap());
+    let server: Server<Filesystem, User> = libunftp::ServerBuilder::with_authenticator(sgen, auth)
+        .pasv_listener(control_sock.local_addr().unwrap().ip())
```

Comment:
Why exactly is it needed to create the passive listener here? Is the IP somehow going to be different from the parent?

Reply:
Prior to this PR, the passive listener socket would be created in the PASV command handler.  However, that doesn't happen until after we enter capsicum mode, and you can't create new sockets once already in capsicum mode.  So we need to preallocate here.

Reply:
OK Makes sense

Reply:
I think we can do it more easily than that.  We don't need to `listen` on the socket early; just create it.  So we can create the socket but only call `listen` during `Pasv::handle_nonproxy_mode`.  Would that suffice?

Reply:
That indeed would reduce the problem. It would still be best if the chosen port is randomly chosen for each pasv invocation. So that the port is not 'owned' by the session which would lead to higher saturation of the "connection pool". And also wouldn't make the next listening port predictable for an attacker who can somehow observe these connections.

Reply:
Hmm.  I can try.  The restriction is that we can't open new sockets once we enter capability mode.  Is it possible to re-bind an already-listening socket to a new port?

Reply:
We should keep in tact the behavior that:

1. Client issues pasv
2. Server picks a random port in the specified range, and starts listening
3. Client issues a data command and connects to passive port
4. Server accepts connection <--- passive port is now free to be listened to again after a next pasv from any other client

Step 4 is important, so that we can keep the capacity. Sometimes we might configure a small pool of ports, so it could be a potential bottleneck.